### ISSUE TITLE
Conversion from xml to dataless drops all station epochs, except earliest

### ISSUE THREAD

**Issue by jholland-usgs** [iris-edu/stationxml-seed-converter#44]
Versions: 2.0.9 and 2.0.10
All files from my test are attached.

Execute
```
java -jar stationxml-seed-converter-2.0.10-SNAPSHOT.jar --input IU.NWAO.xml --output IU.NWAO.dataless
rdseed -s -f IU.NWAO.dataless > nwao.txt
less nwao.txt 
less IU.NWAO.xml
```

Note that the originating IU.NWAO.xml contains BH1 metadata:
```
<Channel code="BH1" locationCode="00" startDate="2018-09-15T00:00:00" restrictedStatus="open">
     <Latitude>-32.9277</Latitude>
     <Longitude>117.239</Longitude>
     <Elevation>278.1</Elevation>
     <Depth>102</Depth>
     <Azimuth>72</Azimuth>
     <Dip>0</Dip>
```

nwao.txt and IU.NWAO.dataless do not contain BH1 metadata.
[droppingAllEpochs.zip](https://github.com/iris-edu/stationxml-seed-converter/files/4228198/droppingAllEpochs.zip)

**Comment by jholland-usgs** [comment]
The problem appears to be in how you create Blockette 11.  

```
0110021001NWAO 000449  
```
Field 5 points to the sequence number of the logical record that corresponds to the station described in field 4. Unfortunately record 449 is near the end of the dataless, so rdseed skips over most of the epochs.

If you read the SEED manual on blockette 11, you should also find that you need to repeat fields 4/5 for each station in the volume.

The SEED manual also states that multiple blockette 50s for a given station should immediately follow each other.

EG
```
blk50 NWAO 1991-1998
blk50 NWAO 1998-2011
blk50 NWAO 2011-2500

Other blks channel, response, etc...

blk50 OTHER station
blk50 OTHER station
Other blks channel, response, etc...

ETC
```

See page 65 in 2.4 seed manual

```
NOTE: If information in the Station Identifier Blockette [50] were to be changed during the time interval of the
volume, additional blockettes with the new information and new effective dates would immediately follow the first
blockette (if there are several changes, there should be additional blockettes).
```

**Comment by rcasey-iris** [comment]
Hi James-

	Thanks for the feedback.  We'll try to get to addressing this issue as soon as is possible.  One of the technical staff for this project is away from the office for the next couple of days.

	Your point is well taken for the Blockette 11 construct.  The six digit number that appears should be the record index number for the blockette 50 for that station.  The rules are not clear on how to handle multiple epochs for the same station, since Blockette 11 does not take that into consideration, but the Blockette 11 page (pg. 39) does indicate that field 3 is the number of blockette 50s to be found in the station header section, also indicating the number of repeating elements in the Blockette 11.

	So following that logic, for three different NWAO epochs, it would be something like:

011[length]003NWAO 000003NWAO 000009NWAO 0000015.....

	You also touched on a concerning point about the latest 2.4 version of the manual, which seems to contain a holdover note from the 1993 edition.  Though the point is less of a concern for us now, we long abandoned support for SEED headers that stacked 50s on top of each other.  It made dealing with changes in channel epochs especially jarring and broke the sense of hierarchy that nested channels afforded.  If there were something to challenge the notion of stacking 50's, it would be the pseudocode on page 20 of the SEED manual:

for each_station do begin
	Start_Station_Header_Records ();
	Write_Station_ID_Blockette_50(); [50] 

	for each_station_comment do

		Write_Station_Comment_Blockette_51(); [51]

   for each_channel do begin

		Write_Channel_ID_Blockette_52(); [52] 

		for each_stage do begin

			if poles_and_zeros then
			Write_Response_Blockette_53(); [53] or Write_Response_Blockette_60(); [60] 

	.....etc.....


So, as in my example above, you'll find multiple epochs for the same station will tend to have a spacing of a few record increments between them.  This way, every 50 is indexed and not redundant.

	Since we're end of life for SEED, it's unlikely that we'll make any changes to the manual, but the FDSN community may decide to generate an errata page to clarify such points as you discovered on page 65.


	-Rob

**Comment by jholland-usgs** [comment]
Thank you for your response.

Thank you specifically for addressing the blockette 11 issue. Your solution repeating the field4/5 per blk50 will probably work, I haven't tested it.
Another possible solution (might be easier to implement, but may be incorrect depending on SEED interpretations) is changing blk11 field 5 to be the first written record for the station (EG 449->003). This does get rdseed to process the whole dataless.  
  
  
I definitely see your point regarding stacked blockette 50s. They are not intuitive and do not nest like other blockettes. I also think unstacked Blk 50s should have been the standard.

However, the code you reference is specifically for Field Volumes (Blk 5).
The network volume (Blk 10) standard is different, see pages 22-25. It does stack blockette 50s.
```
for each_station do begin
    Start_Station_Header_Records();
    for original_and_any_updates do
        Write_Station_ID_Blockette_50(); [50]
    for each_station_comment do
        Write_Station_Comment_Blockette_51(); [51]
    for each_channel do begin
        for original_channel_and_any_updates do begin
            Write_Channel_ID_Blockette_52(); [52]
            for each_stage do begin
```
Glancing through the changes between versions in the manual, I cannot find reference to a change from stacked to unstacked blk 50s.

Perhaps I have missed something. I am open to correction.

If indeed this is a printing error as you suggest, then it is a most grievous and unacceptable error.


Regardless of the SEED manual, as long as the converter generates rdseed, obspy, libmseed, etc...  compatible dataless/xml, I can adjust our end to match it as needed.

Thanks again for looking at this.

**Comment by rcasey-iris** [comment]
Hi James-

	Well, you lawyer-ed me on that one.  It certainly does specify that for the network volume.  It was in the 1993 edition as well.

	In any case, I think we'll have to agree on precedent and have a solution that is most widely useful.  We want to make sure we continue to support the dataless SEED community as we work to make StationXML more ubiquitous over the next few years.

	Getting this Blockette 11 fix in place should hopefully do the trick this time around.

	Regards,

	-Rob

**Comment by timronan** [comment]
This issue seems to be mislabeled. Using the stationxml-seed-converter-2.0.10-SNAPSHOT.jar instance that was included in the zipped file I was able to converter stationxml->dataless seed and not loose any station epochs. 
```
0500103NWAO -32.927700+117.239000+0380.00068000Narrogin, Australia~0013210102011,120,02:00:00.0000

 0500125NWAO -32.927700+117.239000+0380.00029000Narrogin, Australia~0013210102009,221,00:00:00.0000~2011,120,02:00:00.0000

0500125NWAO -32.927700+117.239000+0380.00094000Narrogin, Australia~0013210101998,289,04:00:00.0000~2009,221,00:00:00.0000

 0500125NWAO -32.927700+117.239000+0380.00016000Narrogin,          Australia~0013210101991,329,00:00:00.0000~1998,289,04:00:00.0000

```

The strings above were taken directly from the provided IU.NWAO.dataless  file. Furthermore, this file can be read into pdcc and all of the station epochs are returned. The problem, as stated above, is that blockette 11 is not being correctly written. We are not repeating fields 4 and 5 are instead only writing out F4 and F5 for the last station record in the stationxml file. It seems as though rdseed uses the B11:F5 to find its first B50 to process and then processes the B050's sequentially. Because we are only reporting the last B11:F5 in the converted dataless, rdseed is only processing the last station record. 

Reiterating statements above, the converter must either write B11:F5 associated with the first station or provide it for all of the stations. It seems like the 

`011[length]003NWAO 000003NWAO 000009NWAO 0000015.....`

logic is more robust and should be used if possible. We will work on the code deck and report when we have a patch. Thanks for the catching this bug.

**Comment by timronan** [comment]
This issue should be resolved by commit af9475c4d258747ecc4f66cce3b913ca9edf6f46 in java-4-seed and commit, with test code, fe84e6314bba5017eb56be5200e7c4405e8f6405 in the converter. A new jar has been added to the release page. Please build and test this code. Let us know if it resolves the rdseed issue. We tested the rdseed issue before posting the commits but it is helpful to have external verification. 

Thanks

**Comment by jholland-usgs** [comment]
Thank you for addressing this.

I have tested the latest build from source.
